rust-runtime/aws-smithy-http-server/src/request/request_id.rs

```diff
@@ -150,3 +171,45 @@ impl<Protocol> IntoResponse<Protocol> for MissingServerRequestId {
         internal_server_error()
     }
 }
+
+pin_project_lite::pin_project! {
+    pub struct ServerRequestIdResponseFuture<Fut> {
+        request_id: ServerRequestId,
+        header_key: Option<Option<HeaderName>>,
+        #[pin]
+        fut: Fut,
+    }
+}
+
+impl<Fut> Future for ServerRequestIdResponseFuture<Fut>
+where
+    Fut: TryFuture<Ok = Response<crate::body::BoxBody>>,
+{
+    type Output = Result<Fut::Ok, Fut::Error>;
+
+    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let fut = this.fut;
+        let request_id = this.request_id;
+        fut.try_poll(cx)
+            .map_ok(|mut res| {
+                if let Some(header_key) = this.header_key.take().expect("Futures should not be polled after completion") {
+                    if let Ok(value) = HeaderValue::from_str(&request_id.id.to_string()) {
+                        res.headers_mut().insert(header_key, value);
+                    }
+                }
+                res
+            })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::ServerRequestId;
```

Comment:
Maybe a couple more tests here? We can make use of `tower::util::service_fn` to spin up an easy inner service.